arch/arc/core/arc_connect.c

```diff
@@ -26,8 +26,15 @@ static struct k_spinlock arc_connect_spinlock;
 /* Generate an inter-core interrupt to the target core */
 void z_arc_connect_ici_generate(u32_t core)
 {
+	u32_t ipi_was_pending;
+
 	LOCKED(&arc_connect_spinlock) {
-		z_arc_connect_cmd(ARC_CONNECT_CMD_INTRPT_GENERATE_IRQ, core);
+		z_arc_connect_cmd(ARC_CONNECT_CMD_INTRPT_READ_STATUS, core);
+		ipi_was_pending = z_arc_connect_cmd_readback();
+		if (!ipi_was_pending) {
```

Comment:
This looks super racy to me.  What if the IPI was handled between that line and the next?  The IPI won't be sent, and the kernel code on our core that requested it (e.g. trying to kill a thread) will spin until the next time the target core takes an interrupt or switches out of its thread, which can be a very long time.

Is there a hardware limitation that you can't flag colliding IPIs like this?  If there is, this might need more elaborate code to be 100% sure the IPI gets delivered.

Reply:
make sense, i will think about it